text/0000-freeze-uninit.md

```diff
@@ -0,0 +1,217 @@
+}
+impl<T> *mut T{
+    pub unsafe fn read_freeze(self) -> T;
+}
+
+// in module `core::mem`
+impl<T> MaybeUninit<T>{
+    pub fn freeze(self) -> Self;
+}
+```
+
+`read_freeze` is an operation that loads a number of bytes from (potentially uninitialized) memory, replacing every byte in the source that is uninitialized with an arbitrary-but-initialized value.  
+The function has similar safety constraints to `core::ptr::read`, in that the source pointer must be *dereferenceable* and well aligned, as well as pointing to a valid value of type `T` (after replacing uninitialized bytes).  
+The same-name functions on the raw pointer types are identical to the free function version, and are provided for convience. 
+Only the bytes of the return value are frozen. The bytes behind the pointer argument are unmodified.  
+**The `read_freeze` operation does not freeze any padding bytes of `T` (if any are present), and those are set to uninit after the read as usual.**
+
+`MaybeUninit::freeze` is a safe, by-value version of `read_freeze`. It takes in a `MaybeUninit` and yields an initialized value, which is either exactly `self` if it is already initialized, or some arbitrary value if it is not. 
+
+The result of `core::ptr::read_freeze` or `MaybeUninit::freeze` is not guaranteed to be valid for `T` if `T` has a more complex validity invariant than an array of `u8` (for example, `char`, `bool`, or a reference type). 
+However, the result is guaranteed to be valid for an integer or floating point type, or an aggregate (struct, union, tuple, or array) only containing (recursively) those types. 
+
+Note that frozen bytes are arbitrary, not random. 
+Rust code must not rely on frozen uninitialized bytes (or unfrozen uninitialized bytes) as a source of entropy, only as values that it does not care about and is fine with garbage-but-consistent computational results from. 
+
+For example, the following function:
+```rust
+pub fn gen_random() -> i32{
+    unsafe{MaybeUnit::uninit().freeze().assume_init()}
+}
+```
+
+can be validily optimized to:
+```rust
+pub fn gen_random() -> i32{
+    4 // Compiler chose a constant by fair dice roll
+}
+```
+
+(See also [XKCD 221](https://xkcd.com/221/))
+
+Note that the value `4` was chosen for expository purposes only, and the same optimization could be validly replace by any other constant, or not at all.
+
+## Relationship to `read_volatile`
+
+`read_volatile` and `read_freeze` are unrelated operations (except insofar as they both `read` from a memory location). 
+`read_volatile` performs an observable side effect (that compilers aren't allowed to remove), but will otherwise act (mostly) the same as `read`. `read_volatile` does not freeze bytes read. 
+In contrast, `read_freeze` is not a side effect (thus can be freely optimized by the compiler to anything equivalent).
+
+It is possible in the future that `read_volatile` may carry a guarantee of freezing (non-padding) bytes, but this RFC does not provide that guarantee.
```

Comment:
So, part of the reason that I mentioned `read_volatile` is that in many senses, it *does* freeze the bytes read.

The purpose of `read_volatile` is to prevent the compiler from assuming that the memory location does not change outside the program, e.g. in the case of things like drivers for hardware that uses direct memory access. However, it also assumes implicitly that the value itself is frozen, since it's not able to read an uninitialized value from that location; it wouldn't be useful if it did that, since that means that any read from that location could potentially be UB based upon the decisions made by some external entity, which doesn't make a lot of sense.

Freezing is still useful for reasons like you said-- volatile reads have side effects that are undesired in most cases where freezing is useful. However, I don't think it's right to say that volatile reads don't freeze the data they read.

Reply:
As it currently is, this is not the case. You can observe the difference with miri or msan.

Reply:
Wait, so, what exactly does that mean for embedded contexts and the DMA case? It's just always been UB to volatile read from these addresses, and only adding freeze makes it not UB?

Reply:
Presumably anything outside the AM that's accessing memory outside the AM isn't setting things to uninit.

Reply:
This CE link shows that llvm also doesn't think that `read_volatile` freezes anything: https://godbolt.org/z/6Td6Mhd67

Note that it's discarding the value obtained via `read_volatile` and not writing it back to `x`. If it froze the result, then that would be an invalid transformation, because code could observe that the value was frozen (see my note about MADV_FREE making this observable even at the machine level).

Reply:
`read_volatile` does not currently freeze the data it reads; `MaybeUninit::<u8>::uninit().as_ptr().read_volatile()` reports UB under Miri, as it produces an uninit `u8`.

> However, it also assumes implicitly that the value itself is frozen, [...]

Yes, this is true, `read_volatile` *assumes* that the value in memory is already "frozen" (i.e. initialized for the type being read); it does not do that freezing itself, and thus it is UB if the assumption is violated.

Also, IIUC the "virtual memory" issues referred to elsewhere in the RFC (involving `MADV_DONTNEED`(?)) are not affected by whether a read is volatile or not(?).

Reply:
(Updated link, also showing that the store is tossed at the llvm level, which means it doesn't think that the load volatile/store pair as being able to change the pointed-to value: https://godbolt.org/z/9Tecb6r7r)

Reply:
> `MaybeUninit::<u8>::uninit().as_ptr().read_volatile()` reports UB under Miri, as it produces an uninit u8.

Right, that makes sense, since the pointer itself is an uninitialized value and isn't frozen at all. I'm thinking more of something like, well, whatever the strict-provenance version of this will be:

```rust
static DMA_LOC: *const u64 = std::ptr::without_provenance(0xDEADBEEF);

fn read_dma() -> u64 {
    unsafe { DMA_LOC.read_volatile() }
}
```

Since the pointer *is* defined, it's just not to known program memory. It would be UB if the memory address isn't accessible to the current program, obviously, since that's just a segfault, but I would expect Rust to assume that the bytes of the read *are* frozen in this case, since you can't do a freeze-read and then a volatile-read.

Reply:
> > `MaybeUninit::<u8>::uninit().as_ptr().read_volatile()` reports UB under Miri, as it produces an uninit u8.
> 
> Right, that makes sense, since the pointer itself is an uninitialized value and isn't frozen at all.

it's not the pointer being uninitialized, but the value pointed to..., hence why Miri complains about an uninit u8 and not loading from an uninit pointer.

Reply:
https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=a4377fdb3344d4ec5136953bba9a70b1 for a more spread out version, where miri reports UB on specifically the `read_volatile` line (also, the rest is safe code, so cannot be UB barring compiler/library bugs).

```rust
use std::mem::MaybeUninit;
fn main() {
    let x = MaybeUninit::<u8>::uninit();
    let p = x.as_ptr();
    let v = unsafe { p.read_volatile() };
    dbg!(v);
}
```

Reply:
Right, I completely misread `MaybeUninit::<u8>::uninit().as_ptr()` as `MaybeUninit::<*const u8>::uninit().assume_init()` for some reason, which is why I was confused. The latter would be UB just by itself.

Reply:
> (Updated link, also showing that the store is tossed at the llvm level, which means it doesn't think that the load volatile/store pair as being able to change the pointed-to value: https://godbolt.org/z/9Tecb6r7r)

[Looks like GCC does the same thing, although ICC and MSVC don't.](https://gcc.godbolt.org/z/rabToGfqj)

I find that *very* unfortunate.

What makes volatile volatile is that it pierces through Abstract Machine semantics and is defined in terms of assembly-level semantics instead.  It has to be defined that way; otherwise it wouldn't make sense to say things like "can't be optimized" or "must be a single load/store instruction".  And assembly-level semantics don't have uninitialized bytes.

I'm not saying it's *impossible* to define volatile in a way where it actually can have uninitialized bytes, justifying the behavior observed here.  But it makes the definition more complicated for little benefit.  My ideal definition for volatile is essentially "equivalent to an `asm!` block containing a load/store instruction".  There are a few caveats, maybe.  The compiler is allowed to reorder volatile accesses with non-volatile ones.  The compiler can assume a volatile load/store will actually be executed, unlike an `asm!` block which can be patched at runtime, although I'm not sure that actually justifies any new optimizations.  Regardless, there should only be a few caveats.  It should be basically equivalent to an `asm!` block.

I am going to file a bug report against LLVM to see what they think about this.

Reply:
I think we've tried in the past to make LLVM consider volatile loads as having an implicit freeze, and they wouldn't have it.

But we could codegen `read_volatile` with a `freeze` of the loaded value...

Reply:
> [Looks like GCC does the same thing, although ICC and MSVC don't.](https://gcc.godbolt.org/z/rabToGfqj)

Hang on a second. ICC and MSVC do both the read and the write, and Clang tosses the write but **GCC tosses the read as well**. That's worse than "the same thing as Clang", it's a straight up violation of the popular understanding of volatile in C. You should also be filing a bug report against GCC.

Reply:
(FTR, xlang from lccc does spec that both loads and stores with `volatile` in the access-class do the same thing as `freeze`ing every read/written byte - given gcc's behaviour appears to just be a bug, it seems like we could say Rust volatile loads/stores `freeze` but don't particularily want to extend the scope of the RFC to modify `read_volatile`/`write_volatile`)